server/src/main/java/org/elasticsearch/index/mapper/FieldTypeLookup.java

```diff
@@ -44,12 +45,14 @@ final class FieldTypeLookup {
     ) {
 
         final Map<String, MappedFieldType> fullNameToFieldType = new HashMap<>();
+        final Map<String, String> fullSubfieldNameToParentPath = new HashMap<>();
         final Map<String, DynamicFieldType> dynamicFieldTypes = new HashMap<>();
         final Map<String, Set<String>> fieldToCopiedFields = new HashMap<>();
         for (FieldMapper fieldMapper : fieldMappers) {
             String fieldName = fieldMapper.name();
             MappedFieldType fieldType = fieldMapper.fieldType();
             fullNameToFieldType.put(fieldType.name(), fieldType);
+            fieldMapper.forEach(mapper -> fullSubfieldNameToParentPath.put(mapper.name(), fieldName));
```

Comment:
Note for reviewer: I originally used `fieldMapper.multiFields().forEach()` here, but then path resolution failed for `search_as_you_type` fields, where the source path of subfields is expected to be the parent field's source path.
I switched to relying on `org.elasticsearch.index.mapper.FieldMapper#iterator()` instead, but I am not sure whether this is the desired functionally for all other field mappers that override it.

Reply:
Yes, `search_as_you_type` is a pain as it exposes its extra fields.  Maybe we can add a method to `FieldMapper` called `sourceUsedBy()` which by default returns the combined values of `copy_to` and multifields, and then `search_as_you_type` can override this and add in its extra fields?  And we can use this to populate the `sourceLocations` map.

Reply:
When I found about the `search_as_you_type` problem, I thought of adding an interface method to capture all of these together, but then I found out about the superclass `FieldMapper#iterator()` implementation that does cover all I needed, but my question [above](https://github.com/elastic/elasticsearch/pull/97092#discussion_r1242259492) is whether it actually covers too much now? If so, I'll add a dedicated source-path method as you proposed.

Reply:
There are a few non-object mappers that override `#iterator()` which I think might break this in weird ways (e.g. percolator or parent-join) so probably best to have a dedicated method which we can add tests for etc.

Reply:
I added a dedicated iterator for this purpose. It currently uses the existing multi-fields and search-as-you-type iterators, but I added unit tests to catch changes in the delegated iterators that may break this.

I excluded copy_to fields because they are treated differently when calculating source paths- as opposed to multi-fields, for copy_to fields we want to add the source field path _**as well as the target field path**_.